### PR TITLE
Discourage P2WSH with too big script or stack

### PR DESCRIPTION
**Pull request by jl2012**
Idea by @sipa

This implements a policy limit for P2WSH, with witnessScript <= 3600 bytes, witness stack item size <= 80 bytes, and witness stack items <= 100

3600 bytes witnessScript and 100 stack items are adequate for a n-of-100 multisig using OP_CHECKSIG, OP_ADD, and OP_EQUAL

The max size for ECDSA signature is 73 bytes and nothing should use more than that with the current scripting language.

This is to prevent abuse of witness space, and reduce the risks of DoS attack with some unknown special and big scripts.

### CODE REVIEW

**Comment by petertodd** [comment]
Concept ACK

As this is not consensus, we can always increase the limit later.

File: src/script/interpreter.cpp [hunk 1]
```diff
@@ -1354,6 +1357,16 @@ static bool VerifyWitnessProgram(const CScriptWitness& witness, int witversion,
         return set_success(serror);
     }
 
+    // Policy limit for stack size and stack element size
+    if (flags & SCRIPT_VERIFY_DISCOURAGE_BIG_P2WSH) {
+        if (stack.size() > 100)
+            return set_error(serror, SCRIPT_ERR_DISCOURAGE_BIG_P2WSH);
+        for (unsigned int i = 0; i < stack.size(); i++) {
+            if (stack.at(i).size() > 80)
```
Comment (instagibbs) [inline src/script/interpreter.cpp]: are there really no near/mid-term uses of something larger than 80 bytes?

Comment (sipa) [inline src/script/interpreter.cpp]: Apart from the hashing operations, there is not even an opcode in the scripting language that can process them.

File: src/script/interpreter.cpp [hunk 2]
```diff
@@ -1354,6 +1357,16 @@ static bool VerifyWitnessProgram(const CScriptWitness& witness, int witversion,
         return set_success(serror);
     }
 
+    // Policy limit for stack size and stack element size
```
Comment (instagibbs) [inline src/script/interpreter.cpp]: Might make sense to stick this inside the witversion==0 block if we want to tie both behaviors to specific versions or neither if we don't.

**Comment by sipa** [comment]
I don't think this needs to be implemented as a script interpreter option. It can be done long before script execution is attempted.

**Comment by dcousens** [comment]
> I don't think this needs to be implemented as a script interpreter option. It can be done long before script execution is attempted.

The modularity of the flag is nice though.

**Comment by jl2012** [comment]
@sipa @instagibbs @dcousens Another option is make it inside #8499, which creates a smaller diff and won't touch the consensus critical files

**Comment by sipa** [comment]
@jl2012 Yes, I prefer doing it there.

**Comment by jl2012** [comment]
Implemented in a different way at #8499

**Comment by jl2012** [comment]
Closed in favor of #8499

**Comment by laanwj** [comment]
This was closed unmerged. Removing needs backport tag.

**Pull request closed without merging**